tf-psa-crypto/include/psa/crypto.h

```diff
@@ -4812,6 +4812,301 @@ psa_status_t psa_verify_hash_abort(
+
+/**
+ * \brief                       Start a key generation operation, in an
+ *                              interruptible manner.
+ *
+ * \see                         \c psa_generate_key_iop_complete()
+ *
+ * \warning                     This is a beta API, and thus subject to change
+ *                              at any point. It is not bound by the usual
+ *                              interface stability promises.
+ *
+ * \note                        This function combined with \c
+ *                              psa_generate_key_iop_complete() is equivalent
+ *                              to \c psa_generate_key() but \c
+ *                              psa_generate_key_iop_complete() can return
+ *                              early and resume according to the limit set with
+ *                              \c psa_interruptible_set_max_ops() to reduce the
+ *                              maximum time spent in a function.
+ *
+ * \note                        Users should call
+ *                              \c psa_generate_key_iop_complete() repeatedly
+ *                              on the same operation object after a successful
+ *                              call to this function until \c
+ *                              psa_generate_key_iop_complete() either returns
+ *                              #PSA_SUCCESS or an error.
+ *                              \c psa_generate_key_iop_complete() will return
+ *                              #PSA_OPERATION_INCOMPLETE if there is more work
+ *                              to do. Alternatively users can call
+ *                              \c psa_generate_key_iop_abort() at any point
+ *                              if they no longer want the result.
+ *
+ * \note                        This function clears the number of ops completed
+ *                              as part of the operation. Please ensure you copy
+ *                              this value via
+ *                              \c psa_generate_key_iop_get_num_ops() if
+ *                              required before calling.
+ *
+ * \note                        If this function returns an error status, the
+ *                              operation enters an error state and must be
+ *                              aborted by calling \c
+ *                              psa_generate_key_iop_abort().
+ *
+ * \param[in, out] operation    The \c psa_generate_key_iop_t to use.
+ *                              This must be initialized as per the
+ *                              documentation for
+ *                              \c psa_generate_key_iop_t, and be inactive.
+
+ * \param[in] attributes        The attributes for the new key.
+ *                              The following attributes are required for all
+ *                              keys:
+ *                              * The key type. It must not be an asymmetric
+ *                                public key.
```

Comment:
Maybe we should rather say that only key-pair types are accepted?

Although I suppose it depends how we handle fallback for key types for which we don't support interruptible generation. The simplest implementation is
```
psa_generate_key_iop_setup() {
    memset(iop, 0);
    iop->attributes = *attributes;
    return PSA_SUCCESS;
}
psa_generate_key_iop_complete() {
    if (iop->attributes.type has explicit interruptible support) {
        return explicit_interruptible_support();
    } else {
        psa_generate_key(&iop->attributes, ...);
        memset(iop, 0);
        iop->ops_count = 1;
    }
}
```
Erroring on a symmetric-key type would actually be an extra line of code.

See also https://github.com/ARM-software/psa-api/pull/199/files#r1704375106

Reply:
From the PSA spec perspective it makes sense to allow implementations to limit the key-pair types accepted (as opposed to mandating the restricted set).

From TF PSA Crypto perspective: in theory we could try to save that one line of code, but it seems like premature optimisation to me. Also, this is a predominantly asymmetric crypto feature, it is less confusing/more user friendly if we don't allow asymmetric keys here.

Reply:
I think If we are not going to support interruptible symmetric key generation we should add it as a note because currently it's only stated in a bullet point under the attributes parameter in the setup function and I think is an important information to the user.